core/lib/multivm/src/versions/vm_latest/tests/block_tip.rs

```diff
@@ -0,0 +1,284 @@
+    // 1. Max number of logs.
+    // 2. Lots of small L2->L1 messages / one large L2->L1 message.
+    // 3. Lots of small bytecodes / one large bytecode.
+    // 4. Lots of storage slot updates.
+
+    let max_logs = execute_test(L1MessengerTestData {
+        l2_to_l1_logs: L2ToL1Log::MIN_L2_L1_LOGS_TREE_SIZE,
+        ..Default::default()
+    });
+
+    let max_messages = execute_test(L1MessengerTestData {
+        // Each L2->L1 message is accompanied by a Log, so the max number of pubdata is bound by it
+        messages: vec![vec![0; 0]; MAX_PUBDATA_PER_BLOCK as usize / L2ToL1Log::SERIALIZED_SIZE],
+        ..Default::default()
+    });
+
+    let long_message = execute_test(L1MessengerTestData {
+        // Each L2->L1 message is accompanied by a Log, so the max number of pubdata is bound by it
+        messages: vec![vec![0; MAX_PUBDATA_PER_BLOCK as usize]; 1],
+        ..Default::default()
+    });
+
+    let max_bytecodes = execute_test(L1MessengerTestData {
+        // Each bytecode must be at least 32 bytes long
+        bytecodes: vec![vec![0; 32]; MAX_PUBDATA_PER_BLOCK as usize / 32],
+        ..Default::default()
+    });
+
+    let long_bytecode = execute_test(L1MessengerTestData {
+        // We have to add 48 since a valid bytecode must have an odd number of 32 byte words
+        bytecodes: vec![vec![0; MAX_PUBDATA_PER_BLOCK as usize + 48]; 1],
+        ..Default::default()
+    });
+
+    let lots_of_small_repeated_writes = execute_test(L1MessengerTestData {
+        // In theory each state diff can require only 5 bytes to be published (enum index + 4 bytes for the key)
+        state_diffs: generate_state_diffs(true, true, MAX_PUBDATA_PER_BLOCK as usize / 5),
+        ..Default::default()
+    });
+
+    let lots_of_big_repeated_writes = execute_test(L1MessengerTestData {
+        // Each big write will approximately require 32 bytes to encode
+        state_diffs: generate_state_diffs(true, false, MAX_PUBDATA_PER_BLOCK as usize / 32),
+        ..Default::default()
+    });
+
+    let lots_of_small_initial_writes = execute_test(L1MessengerTestData {
+        // Each initial write will take at least 32 bytes for derived key + 5 bytes for value
+        state_diffs: generate_state_diffs(false, true, MAX_PUBDATA_PER_BLOCK as usize / 37),
+        ..Default::default()
+    });
+
+    let lots_of_large_initial_writes = execute_test(L1MessengerTestData {
+        // Each big write will take at least 32 bytes for derived key + 32 bytes for value
+        state_diffs: generate_state_diffs(false, false, MAX_PUBDATA_PER_BLOCK as usize / 64),
+        ..Default::default()
+    });
+
+    let max_used_gas = vec![
+        max_logs,
+        max_messages,
+        long_message,
+        max_bytecodes,
+        long_bytecode,
+        lots_of_small_repeated_writes,
+        lots_of_big_repeated_writes,
+        lots_of_small_initial_writes,
+        lots_of_large_initial_writes,
+    ]
+    .into_iter()
+    .max()
+    .unwrap();
```

Comment:
why max? is it possible that we will have 2 or more such events?

Reply:
@Deniallugo Each such event consumes max amount of pubdata, so all of them at once is not possible

Reply:
and it's impossible that some weird combination of them will produce more gas consumption, right?

Reply:
@Deniallugo Most likely no as processing each of the four:

- Logs
- Messages
- Bytecodes
- Storage logs

Is done fully independently